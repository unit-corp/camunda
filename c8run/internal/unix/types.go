package unix

type UnixC8Run struct {}
