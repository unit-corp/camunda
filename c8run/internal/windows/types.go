package windows

type WindowsC8Run struct{}
