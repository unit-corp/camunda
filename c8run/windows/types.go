package main

import (
	"os"
	"os/exec"
)

type C8Run interface {
	OpenBrowser(name string)
	GetProcessTree(commandPid int) []*os.Process
	GetVersionCmd(javaBinaryPath string) *exec.Cmd
	GetElasticsearchCmd(elasticsearchVersion string, parentDir string) *exec.Cmd
	GetConnectorsCmd(javaBinary string, parentDir string, camundaVersion string) *exec.Cmd
	GetCamundaCmd(camundaVersion string, parentDir string, extraArgs string) *exec.Cmd
}

type C8RunSettings struct {
	config   string
	detached bool
}
