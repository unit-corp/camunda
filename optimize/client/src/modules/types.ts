/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
 * one or more contributor license agreements. See the NOTICE file distributed
 * with this work for additional information regarding copyright ownership.
 * Licensed under the Camunda License 1.0. You may not use this file
 * except in compliance with the Camunda License 1.0.
 */

import {SerializedEditorState} from 'lexical';

export type GenericEntity<D extends object = Record<string, unknown>> = {
  id: string | null;
  name: string;
  lastModified: string;
  created: string;
  owner: string;
  lastModifier: string;
  currentUserRole: string;
  data: D;
};

export type EntityListEntity<D extends object = Record<string, unknown>> = GenericEntity<D> & {
  id: string;
  entityType: string;
};

type FilterFilterApplicationLevel = 'instance' | 'view';

export type InstanceStateFilterType =
  | 'runningInstancesOnly'
  | 'completedInstancesOnly'
  | 'canceledInstancesOnly'
  | 'nonCanceledInstancesOnly'
  | 'suspendedInstancesOnly'
  | 'nonSuspendedInstancesOnly';

export type FlowNodeStateFilterType =
  | 'runningFlowNodesOnly'
  | 'completedFlowNodesOnly'
  | 'canceledFlowNodesOnly'
  | 'completedOrCanceledFlowNodesOnly';

export type IncidentFilterType =
  | 'includesOpenIncident'
  | 'includesResolvedIncident'
  | 'includesClosedIncident'
  | 'doesNotIncludeIncident';

export type FilterDataProps = {
  flowNodeDuration: Record<string, FilterData>;
  runningInstancesOnly: boolean | null;
  processInstanceDuration: FilterData;
} & {
  [key in 'instanceStartDate' | 'instanceEndDate']: Partial<DateFilterType>;
} & {
  [key in 'flowNodeStartDate' | 'flowNodeEndDate']: Partial<
    DateFilterType & {flowNodeIds: string[] | null}
  >;
} & {
  [key in 'assignee']: {
    values?: (string | null)[];
    operator?: string;
  };
} & {
  [key in InstanceStateFilterType | FlowNodeStateFilterType | IncidentFilterType]:
    | FilterData
    | undefined;
} & {
  [key in 'executedFlowNodes' | 'executingFlowNodes' | 'canceledFlowNodes']: {
    values?: string[];
    operator?: string;
  };
};

export type FilterType = keyof FilterDataProps;

export interface ProcessFilter<T extends FilterType = FilterType> {
  type: T;
  data: FilterDataProps[T];
  filterLevel: FilterFilterApplicationLevel;
  appliedTo: string[];
}

export interface FilterData {
  value: string | number;
  unit: string;
  operator?: string;
}

type VariableType =
  | 'String'
  | 'Short'
  | 'Long'
  | 'Double'
  | 'Integer'
  | 'Boolean'
  | 'Date'
  | 'Object'
  | 'Json';

type ViewProcessViewEntity = 'flowNode' | 'userTask' | 'processInstance' | 'variable' | 'incident';

interface ProcessViewProperty {
  name: string;
  type: VariableType;
}

interface ProcessView {
  entity: ViewProcessViewEntity;
  properties: (ProcessViewProperty | string)[];
}

type GroupByValue = {
  unit?: string;
  type?: string | null;
};

type GroupByType =
  | 'assignee'
  | 'duration'
  | 'endDate'
  | 'flowNodes'
  | 'none'
  | 'runningDate'
  | 'startDate'
  | 'userTasks'
  | 'variable';

interface ProcessGroupBy {
  type: GroupByType;
  value: GroupByValue | null;
}

type DistributedByType =
  | 'assignee'
  | 'endDate'
  | 'flowNode'
  | 'none'
  | 'process'
  | 'startDate'
  | 'userTask'
  | 'variable';

interface DistributedBy {
  type: DistributedByType;
  value: GroupByValue | null;
}

type ProcessReportVisualization = 'number' | 'table' | 'bar' | 'barLine' | 'line' | 'pie' | 'heat';

type AggregationTypeType = 'avg' | 'min' | 'max' | 'sum' | 'percentile';

interface AggregationType<Type extends AggregationTypeType = AggregationTypeType> {
  type: Type;
  value: number | null;
}

type TargetValueUnit =
  | 'millis'
  | 'seconds'
  | 'minutes'
  | 'hours'
  | 'days'
  | 'weeks'
  | 'months'
  | 'years';

type GroupAggregateByDateUnit = 'year' | 'month' | 'week' | 'day' | 'hour' | 'minute' | 'automatic';

interface TableColumns {
  includeNewVariables: boolean;
  excludedColumns: string[];
  includedColumns: string[];
  columnOrder: string[];
}

interface BaseLine {
  unit: TargetValueUnit;
  value: string;
}

interface TargetValue {
  unit: TargetValueUnit;
  value: string;
  isBelow: boolean;
}

interface DurationProgressTargetValue extends TargetValue {
  baseline: BaseLine;
  target: TargetValue;
}

interface CountProgressTargetValue extends TargetValue {
  baseline: string;
  target: string;
}

interface SingleReportTargetValue {
  countChart: TargetValue;
  durationProgress: DurationProgressTargetValue;
  active: boolean;
  countProgress: CountProgressTargetValue;
  durationChart: TargetValue;
  isKpi: boolean;
}

interface HeatmapTargetValueValue {
  unit: TargetValueUnit;
  value: string;
}

interface HeatmapTargetValue {
  active: boolean;
  values: Record<string, HeatmapTargetValueValue>;
}

type CustomBucketUnit =
  | 'year'
  | 'month'
  | 'week'
  | 'day'
  | 'hour'
  | 'minute'
  | 'second'
  | 'millisecond';

interface CustomBucket {
  active: boolean;
  bucketSize: string;
  bucketSizeUnit: CustomBucketUnit;
  baseline: string;
  baselineUnit: CustomBucketUnit;
}

type SortingOrder = 'asc' | 'desc';

interface ReportSorting {
  by?: string | null;
  order?: SortingOrder | null;
}

interface ProcessPart {
  start: string;
  end: string;
}

interface MeasureVisualizations {
  frequency: string;
  duration: string;
}

type UserTaskDurationTimes = 'idle' | 'work' | 'total';

export interface SingleReportConfiguration {
  color: string;
  aggregationTypes: AggregationType[];
  userTaskDurationTimes: UserTaskDurationTimes[];
  showInstanceCount: boolean;
  pointMarkers: boolean;
  precision: number | null;
  hideRelativeValue: boolean;
  hideAbsoluteValue: boolean;
  alwaysShowRelative: boolean;
  alwaysShowAbsolute: boolean;
  showGradientBars: boolean;
  xml: string | null;
  tableColumns: TableColumns;
  targetValue: SingleReportTargetValue;
  heatmapTargetValue: HeatmapTargetValue;
  groupByDateVariableUnit: GroupAggregateByDateUnit;
  distributeByDateVariableUnit: GroupAggregateByDateUnit;
  customBucket: CustomBucket;
  distributeByCustomBucket: CustomBucket;
  sorting?: ReportSorting | null;
  processPart?: ProcessPart | null;
  measureVisualizations: MeasureVisualizations;
  stackedBar: boolean;
  horizontalBar: boolean;
  logScale: boolean;
  yLabel: string;
  xLabel: string;
}

interface SingleReportData {
  configuration: SingleReportConfiguration;
  definitions: Definition[];
}

export interface SingleProcessReportData extends SingleReportData {
  filter: ProcessFilter[];
  view: ProcessView | null;
  groupBy: ProcessGroupBy | null;
  distributedBy: DistributedBy;
  visualization: ProcessReportVisualization | null;
  managementReport: boolean;
  instantPreviewReport: boolean;
  userTaskReport: boolean;
}

export interface SingleProcessReportResultData {
  key: string;
  label: string;
  value: string | number;
}

export type ReportType = 'process';

export interface Report<
  Type extends ReportType = 'process',
  Data extends object = SingleProcessReportData,
  Result = unknown | undefined,
> extends GenericEntity<Data> {
  id: string;
  collectionId?: string | null;
  reportType: Type;
  description: string | null;
  result: Result;
}

export type SingleProcessReport = Report<
  'process',
  SingleProcessReportData,
  {data: SingleProcessReportResultData[]}
>;

export type GenericReport = SingleProcessReport;

type DashboardTileCommonProps = {
  id: string;
  position: {
    x: number;
    y: number;
  };
  dimensions: {
    width: number;
    height: number;
  };
};

export interface OptimizeReportTile extends DashboardTileCommonProps {
  type: 'optimize_report';
  configuration: GenericReport['data']['configuration'];
}

export interface ExternalTile extends DashboardTileCommonProps {
  type: 'external_url';
  configuration: {external: string};
}

export interface TextTile extends DashboardTileCommonProps {
  type: 'text';
  configuration: {text: SerializedEditorState | null};
}

export type DashboardTile = OptimizeReportTile | ExternalTile | TextTile;

export interface Definition {
  identifier: string;
  displayName?: string | JSX.Element[];
  name?: string;
  key?: string;
  tenantIds?: (string | null)[];
  versions?: string[];
  flowNodeIds?: string[];
  type?: string;
}

export type Tenant = {id: string | null; name?: string};

export type Source = {
  definitionKey: string;
  definitionType?: string;
  tenants: (string | null)[];
};

export type Variable = {id?: string; name: string; type: string; label?: string | null};

type CommonFilter = {includeUndefined?: boolean; excludeUndefined?: boolean; name?: string};

type FixedFilter = {
  type: 'fixed';
  start: string | null;
  end: string | null;
};

type RollingFilter = {
  type: 'rolling' | 'custom';
  start: {value?: number | string; unit?: string} | null;
  end: string | null;
  customNum?: string;
};

type OtherFilter = {
  type: '' | 'relative' | 'today' | 'yesterday' | 'this' | 'last' | 'between' | 'after' | 'before';
  start: {value?: number | string; unit?: string} | null;
  end: string | null;
};

export type DateFilterType = CommonFilter & (FixedFilter | RollingFilter | OtherFilter);

interface CommonFilterState {
  type: string;
  unit: string;
  valid?: boolean;
  startDate: Date | null;
  endDate: Date | null;
  includeUndefined?: boolean;
  excludeUndefined?: boolean;
  applyTo?: Definition[];
  values?: (string | number | boolean | null)[];
  operator?: string;
  customNum: string;
}

export interface NoDateFilterState extends CommonFilterState {
  type: 'this' | 'last' | 'yesterday' | 'today';
  startDate: null;
  endDate: null;
}

export interface BetweenFilterState extends CommonFilterState {
  type: 'between';
  startDate: Date;
  endDate: Date;
}

export interface BeforeFilterState extends CommonFilterState {
  type: 'before';
  startDate: Date | null;
  endDate: Date;
}

export interface AfterFilterState extends CommonFilterState {
  type: 'after';
  startDate: Date;
  endDate: Date | null;
}

export interface CustomFilterState extends CommonFilterState {
  type: 'custom';
}

export type FilterState =
  | CommonFilterState
  | BetweenFilterState
  | BeforeFilterState
  | AfterFilterState
  | CustomFilterState;

export interface AnalysisDurationChartEntry {
  key: number;
  value: number;
  outlier: boolean;
}
