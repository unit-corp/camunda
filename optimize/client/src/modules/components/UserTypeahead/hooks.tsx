import {useCallback, useState} from 'react';

import debouncePromise from 'debouncePromise';

import {Identity, searchIdentities} from './service';

const debounceRequest = debouncePromise();

export function useLoadIdentities({
  excludeGroups,
  fetchUsers,
}: {
  fetchUsers?: (
    query: string,
    excludeGroups?: boolean
  ) => Promise<{total: number; result: Identity[]}>;
  excludeGroups: boolean;
}) {
  const [loading, setLoading] = useState(true);
  const [identities, setIdentities] = useState<Identity[]>([]);
  const loadNewValues = useCallback(
    async (query: string, delay = 0) => {
      setLoading(true);

      const {result} = await debounceRequest(async () => {
        return await (fetchUsers || searchIdentities)(query, excludeGroups);
      }, delay);

      setIdentities(result);
      setLoading(false);
    },
    [fetchUsers, excludeGroups]
  );

  return {loading, setLoading, identities, loadNewValues};
}
