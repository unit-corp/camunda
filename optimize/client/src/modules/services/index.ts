/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
 * one or more contributor license agreements. See the NOTICE file distributed
 * with this work for additional information regarding copyright ownership.
 * Licensed under the Camunda License 1.0. You may not use this file
 * except in compliance with the Camunda License 1.0.
 */

export function getRandomId() {
  return Math.random().toString(36).slice(2);
}

export {
  loadReports,
  loadEntities,
  copyEntity,
  deleteEntity,
  createEntity,
  getEntityIcon,
} from './entityService';
export {UNAUTHORIZED_TENANT_ID} from './tenantService';
export * as formatters from './formatters';
export {loadProcessDefinitionXml, loadVariables} from './dataLoaders';
export {numberParser} from './NumberParser';
export {
  TEXT_REPORT_MAX_CHARACTERS,
  isTextTileTooLong,
  isTextTileValid,
  loadRawData,
  evaluateReport,
} from './reportService';
export {addSources, getCollection} from './collectionService';
export {default as getScreenBounds} from './getScreenBounds';
export {default as ignoreFragments} from './ignoreFragments';
export {default as isReactElement} from './isReactElement';
export {incompatibleFilters} from './incompatibleFilters';
export {loadDefinitions} from './loadDefinitions';

export type {Definition} from './loadDefinitions';
export type {ReportEvaluationPayload} from './reportService';
