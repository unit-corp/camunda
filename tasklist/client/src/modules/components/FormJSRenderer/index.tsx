/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
 * one or more contributor license agreements. See the NOTICE file distributed
 * with this work for additional information regarding copyright ownership.
 * Licensed under the Camunda License 1.0. You may not use this file
 * except in compliance with the Camunda License 1.0.
 */

import {useCallback, useEffect, useRef, useState} from 'react';
import type {Variable} from 'modules/types';
import {FormManager} from 'modules/formManager';
import {mergeVariables} from './mergeVariables';
import {ValidationMessage} from './ValidationMessage';
import {getFieldLabels} from './getFieldLabels';
import {usePrefersReducedMotion} from 'modules/hooks/usePrefersReducedMotion';
import styles from './styles.module.scss';
import '@bpmn-io/form-js-viewer/dist/assets/form-js-base.css';
import '@bpmn-io/form-js-carbon-styles/src/carbon-styles.scss';
import type {FileUploadMetadata} from 'modules/mutations/useUploadDocuments';
import set from 'lodash/set';

type Props = {
  handleSubmit: (variables: Variable[]) => Promise<void>;
  handleFileUpload?: (
    files: Map<string, File[]>,
  ) => Promise<Map<string, FileUploadMetadata[]>>;
  schema: string;
  data?: Record<string, unknown>;
  readOnly?: boolean;
  onMount?: (formManager: FormManager) => void;
  onRender?: () => void;
  onImportError?: () => void;
  onSubmitStart?: () => void;
  onSubmitError?: (error: unknown) => void;
  onSubmitSuccess?: () => void;
  onValidationError?: () => void;
};

function htmlDomId(
  fieldId: string,
  formId?: string,
  indices?: string[],
): string {
  const result = ['fjs-form'];
  if (formId) {
    result.push('-', formId);
  }
  result.push('-', fieldId);
  if (indices) {
    result.push(indices.join(''));
  }
  return result.join('');
}

function useScrollToError(manager: FormManager) {
  const prefersReducedMotion = usePrefersReducedMotion();
  return useCallback(
    (fieldId: string) => {
      if (prefersReducedMotion) {
        return;
      }
      const form = manager.get('form');
      const ffr = manager.get('formFieldRegistry');
      const field = ffr.get(fieldId);
      const indicies: string[] = [];
      if (field._path.length > 2) {
        let parent = field;
        while (parent._path.length > 2) {
          parent = ffr.get(parent._parent);
          if (parent.type === 'dynamiclist') {
            indicies.push('_0');
          }
        }
      }
      let firstInvalidDomId: string | undefined;
      if (field.type === 'radio' || field.type === 'checklist') {
        firstInvalidDomId = htmlDomId(fieldId, form._id, [...indicies, '-0']);
      } else if (field.type === 'datetime') {
        firstInvalidDomId = htmlDomId(fieldId, form._id, [
          ...indicies,
          '-date',
        ]);
      } else {
        firstInvalidDomId = htmlDomId(fieldId, form._id, indicies);
      }
      if (firstInvalidDomId) {
        document
          .getElementById(firstInvalidDomId)
          ?.scrollIntoView({behavior: 'auto', block: 'center'});
      }
    },
    [manager, prefersReducedMotion],
  );
}

function extractFilePath(data: object, path: string = ''): Map<string, string> {
  let paths = new Map();

  if (data === null || data === undefined) {
    return paths;
  }

  for (const [key, value] of Object.entries(data)) {
    const itemPath = path.length === 0 ? key : `${path}.${key}`;

    if (typeof value === 'string' && value.startsWith('files::')) {
      paths.set(value, itemPath);
      continue;
    }

    if (Array.isArray(value)) {
      const result = value
        .map((dynamicListItem, index) =>
          extractFilePath(dynamicListItem, `${itemPath}[${index}]`),
        )
        .reduce((acc, cur) => new Map([...acc, ...cur]), new Map());

      paths = new Map([...paths, ...result]);
      continue;
    }

    if (
      typeof value === 'object' &&
      value !== null &&
      Object.keys(value).length > 0
    ) {
      const result = extractFilePath(
        value,
        path.length === 0 ? key : `${itemPath}.${key}`,
      );
      paths = new Map([...paths, ...result]);
      continue;
    }
  }

  return paths;
}

function injectFileMetadataIntoData(options: {
  data: Record<string, unknown>;
  fileMetadata: Map<string, FileUploadMetadata[]>;
  pathsToInject: Map<string, string>;
}): Record<string, unknown> {
  const {data, fileMetadata, pathsToInject} = options;
  let result = data;

  pathsToInject.forEach((filepickerPath, fileKey) => {
    const metadata = fileMetadata.get(fileKey);

    if (metadata === undefined) {
      return;
    }

    result = set(result, filepickerPath, metadata);
  });

  return result;
}

const FormJSRenderer: React.FC<Props> = ({
  handleSubmit,
  schema,
  data = {},
  readOnly,
  onMount,
  onRender,
  onImportError,
  onSubmitStart,
  onSubmitError,
  onSubmitSuccess,
  onValidationError,
  handleFileUpload = () => Promise.resolve(new Map()),
}) => {
  const formManagerRef = useRef<FormManager>(new FormManager());
  const formContainerRef = useRef<HTMLDivElement | null>(null);
  const [invalidFields, setInvalidFields] = useState<
    {ids: string[]; labels: string[]} | undefined
  >();
  const scrollToError = useScrollToError(formManagerRef.current);

  useEffect(() => {
    const formManager = formManagerRef.current;

    onMount?.(formManager);
  }, [onMount]);

  useEffect(() => {
    function render() {
      const formManager = formManagerRef.current;
      const container = formContainerRef.current;

      if (container !== null) {
        onRender?.();
        formManager.render({
          container,
          schema,
          data,
          onImportError,
          onSubmit: async ({data: newData, errors, files = new Map()}) => {
            onSubmitStart?.();
            setInvalidFields(undefined);
            if (Object.keys(errors).length === 0) {
              const enrichedData =
                files.size === 0
                  ? newData
                  : injectFileMetadataIntoData({
                      data: newData,
                      fileMetadata: await handleFileUpload(files),
                      pathsToInject: extractFilePath(newData),
                    });
              const variables = Object.entries(
                mergeVariables(data, enrichedData),
              ).map(
                ([name, value]) =>
                  ({
                    name,
                    value: JSON.stringify(value),
                  }) as Variable,
              );

              try {
                await handleSubmit(variables);
                onSubmitSuccess?.();
              } catch (error) {
                onSubmitError?.(error);
              }
            } else {
              onValidationError?.();
              const fieldIds = Object.keys(errors);
              setInvalidFields({
                ids: fieldIds,
                labels: getFieldLabels(formManager, fieldIds),
              });
              if (fieldIds.length > 0) {
                scrollToError(fieldIds[0]);
              }
              return;
            }
          },
        });
      }
    }

    render();
  }, [
    schema,
    handleSubmit,
    handleFileUpload,
    onRender,
    onImportError,
    onSubmitStart,
    onSubmitSuccess,
    onSubmitError,
    onValidationError,
    data,
    scrollToError,
  ]);

  useEffect(() => {
    const formManager = formManagerRef.current;

    return () => {
      formManager.detach();
    };
  }, []);

  useEffect(() => {
    const formManager = formManagerRef.current;

    formManager.setReadOnly(Boolean(readOnly));
  }, [readOnly]);

  return (
    <>
      <div className={styles.container}>
        <div ref={formContainerRef} className={styles.formRoot} />
      </div>
      {invalidFields !== undefined ? (
        <ValidationMessage
          fieldIds={invalidFields.ids}
          fieldLabels={invalidFields.labels}
        />
      ) : null}
    </>
  );
};

export {FormJSRenderer};
