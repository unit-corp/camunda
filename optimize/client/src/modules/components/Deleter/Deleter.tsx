/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
 * one or more contributor license agreements. See the NOTICE file distributed
 * with this work for additional information regarding copyright ownership.
 * Licensed under the Camunda License 1.0. You may not use this file
 * except in compliance with the Camunda License 1.0.
 */

import {useState, useEffect, useRef} from 'react';
import {Button} from '@carbon/react';

import {Modal, Loading} from 'components';
import {showError} from 'notifications';
import {deleteEntity as deleteEntityService} from 'services';
import {t} from 'translation';
import {useErrorHandling} from 'hooks';
import {EntityListEntity} from 'types';
import {ErrorResponse} from 'request';

const sectionOrder = ['report', 'dashboard', 'alert', 'collection'];

interface DeleterProps {
  entity: EntityListEntity;
  type: string;
  descriptionText?: string;
  deleteText?: string;
  deleteButtonText?: string;
  isReversableAction?: boolean;
  checkConflicts?: (entity: EntityListEntity) => Promise<{conflictedItems: []}>;
  onConflict?: () => void;
  onClose: () => void;
  onDelete: () => void;
  deleteEntity?: ({entityType, id}: {entityType: string; id: string}) => Promise<Response>;
  getName?: (entity: EntityListEntity) => string;
}

type Conflict = {id: string; type: string; name: string};

export default function Deleter({
  entity,
  type,
  descriptionText,
  deleteText,
  deleteButtonText,
  isReversableAction = true,
  checkConflicts,
  onConflict,
  onClose,
  onDelete,
  deleteEntity = ({entityType, id}) => deleteEntityService(entityType, id),
  getName = ({name}) => name,
}: DeleterProps) {
  const [conflicts, setConflicts] = useState<Record<string, Conflict[]>>({});
  const [loading, setLoading] = useState(false);
  const cancelButtonRef = useRef<HTMLButtonElement>(null);
  const {mightFail} = useErrorHandling();

  useEffect(() => {
    if (entity && checkConflicts) {
      setLoading(true);
      mightFail(
        checkConflicts(entity),
        (response) => {
          if (typeof response === 'boolean') {
            if (response) {
              onConflict?.();
            }
          } else {
            setConflicts(
              response.conflictedItems.reduce(
                (obj: Record<string, Conflict[]>, conflict: Conflict) => {
                  obj[conflict.type] = obj[conflict.type] || [];
                  obj[conflict.type]?.push(conflict);
                  return obj;
                },
                {}
              )
            );
          }
        },
        (error: ErrorResponse) => {
          showError(error);
          setConflicts({});
        },
        () => setLoading(false)
      );
    } else {
      setConflicts({});
      setLoading(false);
      cancelButtonRef.current?.focus();
    }
  }, [entity, checkConflicts, mightFail, onConflict]);

  useEffect(() => {
    if (!loading) {
      cancelButtonRef.current?.focus();
    }
  }, [loading]);

  const handleDelete = () => {
    if (!entity) {
      return;
    }

    setLoading(true);
    mightFail(
      deleteEntity(entity),
      () => {
        onDelete();
        handleClose();
      },
      showError,
      () => setLoading(false)
    );
  };

  const handleClose = () => {
    setConflicts({});
    setLoading(false);
    onClose();
  };

  if (!entity) {
    return null;
  }

  const translatedType = t(`common.deleter.types.${type}`).toString();

  return (
    <Modal open onClose={handleClose} className="Deleter">
      <Modal.Header title={deleteText || t('common.deleteEntity', {entity: translatedType})} />
      <Modal.Content>
        {loading ? (
          <Loading />
        ) : (
          <>
            <p>
              {descriptionText ||
                t('common.deleter.permanent', {
                  name: getName(entity),
                  type: translatedType,
                })}
            </p>
            {Object.keys(conflicts)
              .sort((a, b) => sectionOrder.indexOf(a) - sectionOrder.indexOf(b))
              .map((conflictType) => (
                <div key={conflictType}>
                  {t(`common.deleter.affectedMessage.${type}.${conflictType}`)}
                  <ul>
                    {conflicts[conflictType]?.map(({id, name}) => <li key={id}>{name || id}</li>)}
                  </ul>
                </div>
              ))}
            <p>{!isReversableAction && <b>{t('common.deleter.noUndo')}</b>}</p>
          </>
        )}
      </Modal.Content>
      <Modal.Footer>
        <Button
          disabled={loading}
          className="close"
          onClick={handleClose}
          ref={cancelButtonRef}
          kind="secondary"
        >
          {t('common.cancel')}
        </Button>
        <Button kind="danger" disabled={loading} className="confirm" onClick={handleDelete}>
          {deleteButtonText || deleteText || t('common.deleteEntity', {entity: translatedType})}
        </Button>
      </Modal.Footer>
    </Modal>
  );
}
