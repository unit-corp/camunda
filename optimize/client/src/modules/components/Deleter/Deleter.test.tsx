/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
 * one or more contributor license agreements. See the NOTICE file distributed
 * with this work for additional information regarding copyright ownership.
 * Licensed under the Camunda License 1.0. You may not use this file
 * except in compliance with the Camunda License 1.0.
 */

import {runAllEffects} from '__mocks__/react';
import {shallow} from 'enzyme';

import {Modal} from 'components';
import {showError} from 'notifications';
import {deleteEntity} from 'services';
import {useErrorHandling} from 'hooks';
import {EntityListEntity} from 'types';

import Deleter from './Deleter';

jest.mock('hooks', () => ({
  useErrorHandling: jest
    .fn()
    .mockReturnValue({mightFail: jest.fn().mockImplementation((data, cb) => cb(data))}),
}));

jest.mock('services', () => ({
  ...jest.requireActual('services'),
  deleteEntity: jest.fn(),
}));

jest.mock('notifications', () => ({
  showError: jest.fn(),
}));

const entity = {id: 'entityId', name: 'Doomed Report', entityType: 'report'} as EntityListEntity;

const props = {
  entity,
  type: 'report',
  onClose: jest.fn(),
  onDelete: jest.fn(),
};

beforeEach(() => {
  jest.clearAllMocks();
});

it('should show the confirmation modal when entity is set', () => {
  const node = shallow(<Deleter {...props} />);

  expect(node).toMatchSnapshot();
});

it('should allow to check for and display conflicts', () => {
  const conflictChecker = jest
    .fn()
    .mockReturnValue({conflictedItems: [{id: 'conflict1', type: 'dashboard', name: 'conflict1'}]});

  shallow(<Deleter {...props} checkConflicts={conflictChecker} />);
  runAllEffects();

  expect(conflictChecker).toHaveBeenCalled();
});

it('should show an error message if conflict checking goes wrong', async () => {
  const conflictChecker = jest.fn().mockReturnValue({conflictedItems: ['conflict1', 'conflict2']});
  (useErrorHandling as jest.Mock).mockReturnValueOnce({
    mightFail: (_req: unknown, _res: unknown, err: (err: string) => void) =>
      err('everything broke'),
  });

  shallow(<Deleter {...props} checkConflicts={conflictChecker} />);
  runAllEffects();

  expect(showError).toHaveBeenCalledWith('everything broke');
});

it('should delete the entity', () => {
  const node = shallow(<Deleter {...props} />);

  node.find(Modal).find('.confirm').simulate('click');

  expect(deleteEntity).toHaveBeenCalledWith(entity.entityType, entity.id);
});

it('should accept a custom delete executor', () => {
  const spy = jest.fn();
  (deleteEntity as jest.Mock).mockClear();

  const node = shallow(<Deleter {...props} deleteEntity={spy} />);

  node.find(Modal).find('.confirm').simulate('click');

  expect(spy).toHaveBeenCalledWith(entity);
  expect(deleteEntity).not.toHaveBeenCalled();
});

it('should show an error message if deletion goes wrong', () => {
  (useErrorHandling as jest.Mock).mockReturnValueOnce({
    mightFail: (_req: unknown, _res: unknown, err: (err: string) => void) => err('Deleting failed'),
  });

  const node = shallow(<Deleter {...props} />);

  node.find(Modal).find('.confirm').simulate('click');

  expect(showError).toHaveBeenCalledWith('Deleting failed');
});

it('should call the close handler', async () => {
  const spy = jest.fn();
  const node = shallow(<Deleter {...props} onClose={spy} />);
  runAllEffects();

  node.find(Modal).find('.confirm').simulate('click');

  expect(spy).toHaveBeenCalled();
});

it('should accept a custom name formatter', () => {
  const node = shallow(<Deleter {...props} getName={() => 'cool name'} />);

  expect(node.find(Modal.Content)).toMatchSnapshot();
});

it('should use provided delete text for title and button', () => {
  const node = shallow(<Deleter {...props} deleteText="Remove entity" />);

  expect(node).toMatchSnapshot();
});

it('should use the provided delete button text', () => {
  const node = shallow(<Deleter {...props} deleteText="Remove entity" deleteButtonText="Delete" />);

  expect(node.find('.confirm')).toIncludeText('Delete');
});

it('should invoke onConflict if the conflict response is a boolean true', () => {
  const conflictChecker = jest.fn().mockReturnValue(true);
  const spy = jest.fn();

  shallow(<Deleter {...props} onConflict={spy} checkConflicts={conflictChecker} />);
  runAllEffects();

  expect(spy).toHaveBeenCalled();
});

it('should not show the undo warning if the reversableAction prop is set to true', () => {
  const node = shallow(<Deleter {...props} isReversableAction />);
  expect(node.find(Modal.Content).text()).not.toMatch('This action cannot be undone.');
});
