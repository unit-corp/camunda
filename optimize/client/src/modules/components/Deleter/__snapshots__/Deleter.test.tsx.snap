// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should accept a custom name formatter 1`] = `
<ForwardRef>
  <p>
    The 'cool name' Report will be permanently deleted.
  </p>
  <p />
</ForwardRef>
`;

exports[`should show the confirmation modal when entity is set 1`] = `
<Modal
  className="Deleter"
  onClose={[Function]}
  open={true}
>
  <ForwardRef(ModalHeader)
    title="Delete Report"
  />
  <ForwardRef>
    <p>
      The 'Doomed Report' Report will be permanently deleted.
    </p>
    <p />
  </ForwardRef>
  <ForwardRef(ModalFooter)>
    <Button
      className="close"
      disabled={false}
      kind="secondary"
      onClick={[Function]}
    >
      Cancel
    </Button>
    <Button
      className="confirm"
      disabled={false}
      kind="danger"
      onClick={[Function]}
    >
      Delete Report
    </Button>
  </ForwardRef(ModalFooter)>
</Modal>
`;

exports[`should use provided delete text for title and button 1`] = `
<Modal
  className="Deleter"
  onClose={[Function]}
  open={true}
>
  <ForwardRef(ModalHeader)
    title="Remove entity"
  />
  <ForwardRef>
    <p>
      The 'Doomed Report' Report will be permanently deleted.
    </p>
    <p />
  </ForwardRef>
  <ForwardRef(ModalFooter)>
    <Button
      className="close"
      disabled={false}
      kind="secondary"
      onClick={[Function]}
    >
      Cancel
    </Button>
    <Button
      className="confirm"
      disabled={false}
      kind="danger"
      onClick={[Function]}
    >
      Remove entity
    </Button>
  </ForwardRef(ModalFooter)>
</Modal>
`;
